### ISSUE TITLE
[rllib] VTrace in multi-agent non-recurrent setting does not correctly cut/mask batches at episode boundary.

### ISSUE THREAD

**Issue by Bam4d** [ray-project/ray#14128]
### What is the problem?

Ray version: 1.1.0
Python: 3.8
OS: Ubuntu 20.04
Pytorch: 1.7.1

When using IMPALA to train multiple agents i'm coming across numerical instabilities due to batches of zeros polluting the training data. I believe this is due to a padding operation `pad_batch_to_sequences_of_same_size` not being handled correctly in `vtrace_torch_policy.py`

The error I believe is one of (or maybe both of): 

## in the `make_time_major` function here:
https://github.com/ray-project/ray/blob/master/rllib/agents/impala/vtrace_torch_policy.py#L186

###  Reasoning:
  * In MA settings, the batches are padded to equal lengths (#3289).
  * `make_time_major` *should* be cutting at the episode boundary?
  * `make_time_major` does not cut at the episode boundary in seq_lens unless `policy.is_recurrent() == True`?
  * Batches are polluted with zero-valued obs/actions etc

## In the `mask` calculation here:
https://github.com/ray-project/ray/blob/master/rllib/agents/impala/vtrace_torch_policy.py#L146

### Reasoning:
* In MA settings, the batches are padded to equal lengths (#3289).
* Anything beyond episode boundaries should be masked to stop it being being used in gradient calc
* Due to `policy.is_recurrent() == True` being the check, `mask = torch.ones(....)`. No masking applied
* Batches are polluted with zero-valued obs/actions etc

### Reproduction 

```
import os
import sys

import ray
from ray import tune
from ray.rllib.agents.impala import ImpalaTrainer
from ray.rllib.examples.env.rock_paper_scissors import RockPaperScissors

if __name__ == '__main__':
    sep = os.pathsep
    os.environ['PYTHONPATH'] = sep.join(sys.path)

    ray.init(num_gpus=1, local_mode=True)

    env_name = 'ray-reproduce-test'

    config = {
        'framework': 'torch',
        'num_workers': 8,

        # In RockPaperScissors the episodes are always length 10, so we force the padding to rollout 20 to emulate
        # environments that have episode boundaries within the padded rollout
        'rollout_fragment_length':  20,
        'num_envs_per_worker': 4,
        'env': RockPaperScissors,
    }

    stop = {
        # 'training_iteration': 100,
        'timesteps_total': 200000,
    }

    result = tune.run(ImpalaTrainer, config=config, stop=stop)
```

There are no errors thrown or numerical instability in this particular example (as the example is very simple). However if a breakpoint is placed in the `__init__`
method of `VTraceLoss` it can be clearly seen that padded examples are being passed into this calculation and masks are not being applied to stop these padded tensors being part of the calculation.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by Bam4d** [comment]
My current hacky workaround is to override `vtrace_torch_policy.py` and calculate the masks in the following way:

```
    if 'seq_lens' in train_batch:
        max_seq_len = policy.config['rollout_fragment_length']
        mask_orig = sequence_mask(train_batch["seq_lens"], max_seq_len)
        mask = torch.reshape(mask_orig, [-1])
    else:
        mask = torch.ones_like(rewards)
```

`make_time_major` is left unchanged, but I'm unsure if it is working as wanted. Also my hack breaks compatibility with recurrent policies :(
